op-node/rollup/attributes/engine_consolidate.go

```diff
@@ -82,31 +82,41 @@ func checkParentBeaconBlockRootMatch(attrRoot, blockRoot *common.Hash) error {
 	return nil
 }
 
-func checkWithdrawalsMatch(attrWithdrawals *types.Withdrawals, blockWithdrawals *types.Withdrawals) error {
-	if attrWithdrawals == nil && blockWithdrawals == nil {
-		return nil
+// checkWithdrawals checks if the withdrawals list and withdrawalsRoot are as expected in the attributes and block,
+// based on the active hard fork.
+func checkWithdrawals(rollupCfg *rollup.Config, attrs *eth.PayloadAttributes, block *eth.ExecutionPayload) error {
+	if attrs == nil || block == nil {
+		return fmt.Errorf("nil attributes or block")
 	}
 
-	if attrWithdrawals == nil && blockWithdrawals != nil {
-		return fmt.Errorf("expected withdrawals in block to be nil, actual %v", *blockWithdrawals)
-	}
-
-	if attrWithdrawals != nil && blockWithdrawals == nil {
-		return fmt.Errorf("expected withdrawals in block to be non-nil %v, actual nil", *attrWithdrawals)
-	}
+	attrWithdrawals := attrs.Withdrawals
+	blockWithdrawals := block.Withdrawals
 
-	if len(*attrWithdrawals) != len(*blockWithdrawals) {
-		return fmt.Errorf("expected withdrawals in block to be %d, actual %d", len(*attrWithdrawals), len(*blockWithdrawals))
+	if rollupCfg.IsCanyon(uint64(block.Timestamp)) {
+		// canyon is active, we should have an empty withdrawals list
+		if !(blockWithdrawals != nil && len(*blockWithdrawals) == 0) {
+			return fmt.Errorf("canyon: expected withdrawals in block to be non-nil and empty, actual %v", *blockWithdrawals)
+		}
+		if !(attrWithdrawals != nil && len(*attrWithdrawals) == 0) {
+			return fmt.Errorf("canyon: expected withdrawals in attributes to be non-nil and empty, actual %v", *attrWithdrawals)
+		}
+	} else {
+		if attrWithdrawals != nil {
+			return fmt.Errorf("pre-canyon: expected withdrawals in attributes to be nil, actual %v", *attrWithdrawals)
+		}
 	}
 
-	for idx, expected := range *attrWithdrawals {
-		actual := (*blockWithdrawals)[idx]
-
-		if *expected != *actual {
-			return fmt.Errorf("expected withdrawal %d to be %v, actual %v", idx, expected, actual)
+	if rollupCfg.IsIsthmus(uint64(block.Timestamp)) {
+		if block.WithdrawalsRoot == nil {
+			// isthmus is active, we should have a non-nil withdrawalsRoot
+			return fmt.Errorf("isthmus: expected withdrawalsRoot in block to be non-nil")
+		}
+	} else {
+		if !(block.WithdrawalsRoot == nil || *block.WithdrawalsRoot == types.EmptyWithdrawalsHash) {
```

Comment:
If the `withdrawalsRoot` is non-nil and `withdrawals` is nil (not merely empty), then this doesn't return an error. Is that allowed pre-isthmus? Doesn't seem to me that the spec should allow this.

Reply:
Good catch. Looking at the specs: https://github.com/ethereum-optimism/specs/blob/f59de601557a03e31a71e80e784adb27670732ec/specs/protocol/derivation.md#building-individual-payload-attributes

if post canyon, withdrawals must be non nil. Looks like that was never added here

Reply:
Can confirm locally. Though the returned error statement on L98 panics due to a nil pointer deref for this case. Should fix that.